From 664111f178bc071a92387776ccb35af2d4d9989a Mon Sep 17 00:00:00 2001
From: skyjet18 <57456827+skyjet18@users.noreply.github.com>
Date: Sat, 18 Jan 2025 19:41:09 +0100
Subject: [PATCH 18/33] webvttdec: option to insert fake subtitle at the end

Tested on MPEG-DASH with single segments:
- it is not possible to seek in stream after last subtitle
- subtitle stream returns EOF and playback will stop
- happens with exteplayer3
- this option enables inserting fake subtitle at the end with INT_MAX pts
- this enforces, that subtitle stream never ends before audio/video stream
---
 libavformat/webvttdec.c | 23 +++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/libavformat/webvttdec.c b/libavformat/webvttdec.c
index 64a8cda584..7765d66d04 100644
--- a/libavformat/webvttdec.c
+++ b/libavformat/webvttdec.c
@@ -36,6 +36,7 @@ typedef struct {
     FFDemuxSubtitlesQueue q;
     int kind;
     int prefer_hls_mpegts_pts;
+    int fake_last_subtitle;
 } WebVTTContext;
 
 static int webvtt_probe(const AVProbeData *p)
@@ -71,7 +72,8 @@ static int webvtt_read_header(AVFormatContext *s)
     AVStream *st = avformat_new_stream(s, NULL);
     int has_hls_timestamp_map = 0;
     int64_t hls_ts_offset;
-
+    int64_t pos = 0;
+    AVPacket *sub;
 
     if (!st)
         return AVERROR(ENOMEM);
@@ -84,8 +86,7 @@ static int webvtt_read_header(AVFormatContext *s)
 
     for (;;) {
         int i;
-        int64_t pos;
-        AVPacket *sub;
+
         const char *p, *identifier, *settings;
         size_t identifier_len, settings_len;
         int64_t ts_start, ts_end;
@@ -212,6 +213,19 @@ static int webvtt_read_header(AVFormatContext *s)
         SET_SIDE_DATA(settings,   AV_PKT_DATA_WEBVTT_SETTINGS);
     }
 
+    if( webvtt->fake_last_subtitle )
+    {
+        /* insert fake end packet to prevent EOF before video/audio stream ends */
+        sub = ff_subtitles_queue_insert(&webvtt->q, "", 0, 0);
+        if (!sub) {
+            res = AVERROR(ENOMEM);
+            goto end;
+        }
+        sub->pos = pos + INT64_C(1);
+        sub->pts = INT64_MAX - INT64_C(2);
+        sub->duration = INT64_C(1);
+    }
+
     ff_subtitles_queue_finalize(s, &webvtt->q);
 
 end:
@@ -249,7 +263,8 @@ static const AVOption options[] = {
         { "captions",     "WebVTT captions kind",     0, AV_OPT_TYPE_CONST, { .i64 = AV_DISPOSITION_CAPTIONS },     INT_MIN, INT_MAX, KIND_FLAGS, "webvtt_kind" },
         { "descriptions", "WebVTT descriptions kind", 0, AV_OPT_TYPE_CONST, { .i64 = AV_DISPOSITION_DESCRIPTIONS }, INT_MIN, INT_MAX, KIND_FLAGS, "webvtt_kind" },
         { "metadata",     "WebVTT metadata kind",     0, AV_OPT_TYPE_CONST, { .i64 = AV_DISPOSITION_METADATA },     INT_MIN, INT_MAX, KIND_FLAGS, "webvtt_kind" },
-        { "prefer_hls_mpegts_pts", "Use WebVTT embedded HLS MPEGTS timestamps if available.", OFFSET(prefer_hls_mpegts_pts), AV_OPT_TYPE_INT, { .i64 = 1 }, 0, 1, AV_OPT_FLAG_SUBTITLE_PARAM | AV_OPT_FLAG_DECODING_PARAM },
+        { "prefer_hls_mpegts_pts", "Use WebVTT embedded HLS MPEGTS timestamps if available.", OFFSET(prefer_hls_mpegts_pts), AV_OPT_TYPE_INT, { .i64 = 1 }, 0, 1, KIND_FLAGS },
+        { "fake_last_subtitle", "Insert fake last subtitle to prevent subtitle stream end befor audio/video", OFFSET(fake_last_subtitle), AV_OPT_TYPE_INT, { .i64 = 0 }, 0, 1, KIND_FLAGS },
     { NULL }
 };
 
-- 
2.45.2.windows.1

